src/Servers/Kestrel/Transport.Sockets/src/Internal/IOQueue.cs

```diff
@@ -22,35 +21,66 @@ public override void Schedule(Action<object> action, object state)
                 State = state
             };
 
+            // Order is important here with Execute().
+
+            // 1. Enqueue prior to checking _doingWork.
             _workItems.Enqueue(work);
 
-            lock (_workSync)
+            // 2. MemoryBarrier to ensure ordering of Write (workItems.Enqueue) -> Read (_doingWork) is preserved,
+            // as order is reversed between Schedule and Execute, and they are two different memory locations.
+            Thread.MemoryBarrier();
```

Comment:
Do you actually need a full barrier here, or would it be sufficient to make _doingWork volatile?

Reply:
Was experiencing what seemed to be a non-complete race with just volatile in CI; could have been an unrelated test issue (e.g. there were a lot of test timeouts that @davidfowl fixed recently due to hitting blocking in `Dispose()`)

However as `_workItems` and `_doingWork` are two different locations and the store needs to happen prior to the load I thought I might be hitting this, so went with the full barrier:

> [8.2.3.4 Loads May Be Reordered with Earlier Stores to Different Locations](https://www.intel.co.uk/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-vol-3a-part-1-manual.pdf) **pdf**
>
> The Intel-64 memory-ordering model allows a load to be reordered with an earlier store to a different location.
However, loads are not reordered with stores to the same location.
>
> The fact that a load may be reordered with an earlier store to a different location is illustrated by the following
example:
>
> Example 8-3. Loads May be Reordered with Older Stores
>
 ```asm
;Processor 0              Processor 1
 mov [ _x], 1             mov [ _y], 1
 mov r1, [ _y]            mov r2, [ _x]
;Initially x = y = 0
;r1 = 0 and r2 = 0 is allowed
```
> At each processor, the load and the store are to different locations and hence may be reordered. Any interleaving
of the operations is thus allowed. One such interleaving has the two loads occurring before the two stores. This
would result in each load returning value 0.

Reply:
Would it not be more efficient then to just do the Interlocked.Exchange without the MemoryBarrier or the "fast" check?

Reply:
(That said, I'd be surprised if the race you saw was due to the above reordering.  Obviously _workItems itself wouldn't be an issue, and Enqueue itself has a barrier, so the issue would need to have manifested because of the read of _doingWork being reordered with some write in Enqueue after its barrier?  Not sure how that would result in complications.)